File: src/parser.py
from parselets import InfixParselet, PrefixOperatorParselet, PrefixParselet

from .token_type import TokenType


class Parser:
    _infix_parselets: dict[TokenType, InfixParselet]
    _prefix_parselets: dict[TokenType, PrefixParselet]

    def __init__(self):
        self._prefix_parselets = {}

    def parse_expression(self):
        token = self.consume()
        prefix = self._prefix_parselets[token.get_type()]

        if prefix is None:
            raise SyntaxError(f"Could not parse '{token.get_text()}'.")

        left = prefix.parse(parser=self, token=token)

        token = self._look_ahead(0)
        infix = self._infix_parselets[token.get_type()]

        # No infix expression at this point, so we're done.
        if infix is None:
            return left

        self.consume()
        return infix.parse(parser=self, left=left, token=token)

    def register(self, token: TokenType, parselet: PrefixParselet | InfixParselet):
        if isinstance(parselet, PrefixParselet):
            self._prefix_parselets[token] = parselet
        elif isinstance(parselet, InfixParselet):
            self._infix_parselets[token] = parselet

    def prefix(self, token: TokenType):
        self.register(token=token, parselet=PrefixOperatorParselet())

File: src/parser.rs
use std::collections::{HashMap, VecDeque};

use crate::lexer::Lexer;
use crate::token::{Token, TokenType};

pub struct Parser {
    tokens: Iterator<Token>, 
    read: VecDecque<Token>,
    infix_parselets: HashMap<TokenType, InfixParselet>,
    prefix_parselets: HashMap<TokenType, PrefixParselet>,
}

impl Parser {
    pub fn new(tokens: Iterator<Token>) {
        Parser {
            tokens: tokens,
            read: VecDeque::new(),
            infix_parselets: HashMap::new(),
            prefix_parselets: HashMap::new(),
        }
    }

    pub fn register(&mut self, token: TokenType, parselet: Parselet) {
        match parselet {
            PrefixParselet => self.prefix_parselets.insert(token, parselet),
            InfixParselet => self.infix_parselets.insert(token, parselet),
        }
    }

    pub fn parse_expression(&mut self, precedence: usize) {
        let mut token = self.consume(); 
        let prefix = self.prefix_parselets.get(&token.token_type);

        match prefix {
            Some => {
                let mut left = prefix.parse(self, token);
                
                while precedence < self.get_precedence() {
                    token = self.consume();

                    let infix = self.infix_parselets.get(&token.token_type);
                    left = infix.parse(self, left, token); 
                }

                return left;
            },
            None => panic!(format!("Could not parse '{}'", token.value)),
        }
    }

    pub fn match(&mut self, expected: TokenType) -> bool {
        let token = self.look_ahead(0);

        if token.type != expected {
            return false;
        }

        self.consume();
        
        true
    }

    pub fn consume(&mut self, expected: Option<TokenType>) -> Token {
        // Make sure we've read the token.
        token = self.look_ahead(0);

        match expected {
            Some(t) => {
                if t.type != expected {
                    panic!(format!("Expected token {} and found {}", expected, token.type))
                }

                return self.read.pop_front(0);
            },
            None => panic!(format!("Expected token {} and found {}", expected, token.type)),
        }
    }

    fn look_ahead(&mut self, distance: usize) {
        // Read in as many as needed.
        while distance >= self.read.len() {
            self.read.push_back(self._tokens.next());
        }

        self.read.get(distance)
    }

    fn get_precedence(&mut self) -> usize {
        let type = self.look_ahead(0).type;
        let parser = self.infix_parselets.get(type);

        match parser {
            Some(p) => p.get_precedence(),
            None => 0,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{Parser, Token, TokenType};
}
